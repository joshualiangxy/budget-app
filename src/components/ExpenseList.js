import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import getVisibleExpenses from "../selectors/expenses";

export const ExpenseList = props => (
    <div>
        {
            props.expenses.length === 0 
                ? <p>No expenses</p> 
                : props.expenses
                    .map(expense => <ExpenseListItem 
                        {...expense}
                        key={expense.id} />)
        }
    </div>
);

const mapStateToProps = state => ({ expenses: getVisibleExpenses(state.expenses, state.filters) });

export default connect(mapStateToProps)(ExpenseList);