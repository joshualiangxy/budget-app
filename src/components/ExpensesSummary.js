import React from 'react';
import numeral from 'numeral';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import getVisibleExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({
  expenseCount,
  expenseWord,
  expenseTotal
}) => (
  <div className="page-header">
    <div className="content-container">
      <h1 className="page-header__title">
        Viewing <span className="bold">{expenseCount}</span> {expenseWord}
        {expenseCount === 0 ? (
          ''
        ) : (
          <span>
            {' '}
            totalling <span className="bold">{expenseTotal}</span>
          </span>
        )}
      </h1>
      <div className="page-header__actions">
        <Link className="button" to="/create">
          Add Expense
        </Link>
      </div>
    </div>
  </div>
);

const mapStateToProps = state => {
  const expenses = getVisibleExpenses(state.expenses, state.filters);
  const expenseCount = expenses.length;
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
  const expenseTotal = numeral(getExpensesTotal(expenses) / 100).format(
    '$0,0.00'
  );
  return { expenseCount, expenseWord, expenseTotal };
};

export default connect(mapStateToProps)(ExpensesSummary);
