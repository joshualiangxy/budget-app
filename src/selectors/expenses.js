import moment from 'moment';

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter(expense => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, 'day')
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, 'day')
        : true;
      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());
      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((expenseOne, expenseTwo) => {
      switch (sortBy) {
        case 'date':
          return expenseTwo.createdAt - expenseOne.createdAt;
        case 'amount':
          return expenseTwo.amount - expenseOne.amount;
        case 'name':
          return expenseOne.description.localeCompare(expenseTwo.description);
      }
    });
};

export default getVisibleExpenses;
