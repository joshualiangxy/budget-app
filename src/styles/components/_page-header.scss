.page-header {
    background: $light-grey;
    margin-bottom: $l-size;
    padding: $l-size 0;
}

.page-header__actions {
    margin-top: $m-size;
}

.page-header__title {
    margin: 0;
    font-weight: 300;
}

.bold {
    font-weight: bold;
}