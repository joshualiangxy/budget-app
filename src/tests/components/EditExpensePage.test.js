import React from "react";
import { shallow } from "enzyme";
import { EditExpensePage } from "../../components/EditExpensePage";
import expenses from "../fixtures/expenses";

let wrapper, history, editExpense, removeExpense;

beforeEach(() => {
    history = { push: jest.fn() };
    editExpense = jest.fn();
    removeExpense = jest.fn();
    wrapper = shallow(<EditExpensePage 
        expense={expenses[0]}
        history={history}
        editExpense={editExpense}
        removeExpense={removeExpense}
    />);
});

test("Should render EditExpensePage",
    () => {
        expect(wrapper).toMatchSnapshot();
    });

test("Should handle editExpense",
    () => {
        wrapper.find("ExpenseForm").prop("onSubmit")(expenses[1])
        expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[1]);
        expect(history.push).toHaveBeenLastCalledWith('/');
    });

test("Should handle removeExpense",
    () => {
        wrapper.find("button").simulate("click");
        expect(removeExpense).toHaveBeenLastCalledWith(expenses[0].id);
        expect(history.push).toHaveBeenLastCalledWith('/');
    });