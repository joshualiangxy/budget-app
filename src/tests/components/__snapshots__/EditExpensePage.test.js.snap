// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render EditExpensePage 1`] = `
<div>
  <h2>
    Edit Expense
  </h2>
  <ExpenseForm
    expense={
      Object {
        "amount": 9,
        "createdAt": 0,
        "description": "Electricity Bills",
        "id": "1",
        "note": "electric",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
